<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Zhaolian Lu, Keenan Berry, Zhenbin Hu, Yu Zhan, Tae-Hyuk Ahn, Zhenguo Lin" />

<meta name="date" content="2021-08-09" />

<title>TSSr Vignette</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">TSSr Vignette</h1>
<h4 class="author">Zhaolian Lu, Keenan Berry, Zhenbin Hu, Yu Zhan, Tae-Hyuk Ahn, Zhenguo Lin</h4>
<h4 class="date">2021-08-09</h4>



<p>Alternatively, you can install the development version directly from GitHub using <code>devtools</code>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;Linlab-slu/TSSr&quot;</span>)</span></code></pre></div>
<p>And load <code>TSSr</code>:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">library</span>(TSSr)</span></code></pre></div>
<div id="citation" class="section level1">
<h1>Citation</h1>
<p>If you use TSSr, please cite the following article:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw">citation</span>(<span class="st">&quot;TSSr&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="co">#&gt; </span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co">#&gt; To cite package &#39;TSSr&#39; in publications use:</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co">#&gt; </span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">#&gt;   Zhaolian Lu, Keenan Berry, Zhenbin Hu, Yu Zhan, Tae-Hyuk (Ted) Ahn</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co">#&gt;   and Zhenguo Lu (2021) TSS sequencing data analysis. R package version</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co">#&gt;   0.99.0. URL https://github.com/Linlab-slu/TSSr</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co">#&gt; </span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co">#&gt; A BibTeX entry for LaTeX users is</span></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="co">#&gt; </span></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="co">#&gt;   @Manual{,</span></span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="co">#&gt;     title = {{TSSr}: TSS sequencing data analysis},</span></span>
<span id="cb3-13"><a href="#cb3-13"></a><span class="co">#&gt;     author = {Zhaolian Lu and Keenan Berry and Zhenbin Hu and Yu Zhan and Tae-Hyuk Ahn and Zhenguo Lin},</span></span>
<span id="cb3-14"><a href="#cb3-14"></a><span class="co">#&gt;     year = {2021},</span></span>
<span id="cb3-15"><a href="#cb3-15"></a><span class="co">#&gt;     version = {0.99.0},</span></span>
<span id="cb3-16"><a href="#cb3-16"></a><span class="co">#&gt;     url = {https://github.com/Linlab-slu/TSSr},</span></span>
<span id="cb3-17"><a href="#cb3-17"></a><span class="co">#&gt;   }</span></span></code></pre></div>
</div>
<div id="getting-help" class="section level1">
<h1>Getting help</h1>
<p>For general questions about the usage of TSSr, use the <a href="https://support.bioconductor.org">official Bioconductor support forum</a> and tag your question “TSSr”. We strive to answer questions as quickly as possible.</p>
<p>For technical questions, bug reports and suggestions for new features, we refer to the <a href="https://github.com/Linlab-slu/TSSr/issues">TSSr github page</a>.</p>
</div>
<div id="quick-start" class="section level1">
<h1>Quick start</h1>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># Load the example data</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="kw">data</span>(<span class="st">&quot;exampleTSSr&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3"></a>myTSSr &lt;-<span class="st"> </span>exampleTSSr</span></code></pre></div>
<p>Or create a new TSSr object</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># Provide bam files</span></span>
<span id="cb5-2"><a href="#cb5-2"></a>inputFiles &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;S01.sorted.bam&quot;</span>, <span class="st">&quot;S02.sorted.bam&quot;</span>, <span class="st">&quot;S03.sorted.bam&quot;</span>, <span class="st">&quot;S04.sorted.bam&quot;</span>)</span>
<span id="cb5-3"><a href="#cb5-3"></a>myTSSr &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;TSSr&quot;</span>, <span class="dt">genomeName =</span> <span class="st">&quot;BSgenome.Scerevisiae.UCSC.sacCer3&quot;</span></span>
<span id="cb5-4"><a href="#cb5-4"></a>          ,<span class="dt">inputFiles =</span> inputFiles</span>
<span id="cb5-5"><a href="#cb5-5"></a>          ,<span class="dt">inputFilesType=</span> <span class="st">&quot;bam&quot;</span></span>
<span id="cb5-6"><a href="#cb5-6"></a>          ,<span class="dt">sampleLabels =</span> <span class="kw">c</span>(<span class="st">&quot;SL01&quot;</span>,<span class="st">&quot;SL02&quot;</span>,<span class="st">&quot;SL03&quot;</span>,<span class="st">&quot;SL04&quot;</span>)</span>
<span id="cb5-7"><a href="#cb5-7"></a>          ,<span class="dt">sampleLabelsMerged =</span> <span class="kw">c</span>(<span class="st">&quot;control&quot;</span>,<span class="st">&quot;treat&quot;</span>)</span>
<span id="cb5-8"><a href="#cb5-8"></a>          ,<span class="dt">mergeIndex =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>)</span>
<span id="cb5-9"><a href="#cb5-9"></a>          ,<span class="dt">refSource =</span> <span class="st">&quot;saccharomyces_cerevisiae.SGD.gff&quot;</span></span>
<span id="cb5-10"><a href="#cb5-10"></a>          ,<span class="dt">organismName =</span> <span class="st">&quot;saccharomyces cerevisiae&quot;</span>)</span>
<span id="cb5-11"><a href="#cb5-11"></a>myTSSr</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="co"># Get TSS</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="kw">getTSS</span>(myTSSr)</span></code></pre></div>
<p>TSS data processing</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># Merge replicates</span></span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="kw">mergeSamples</span>(myTSSr)</span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="co"># Normalization</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="kw">normalizeTSS</span>(myTSSr)</span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="co">#&gt; </span></span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co">#&gt; Normalizing TSS matrix...</span></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co"># TSS filtering</span></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="kw">filterTSS</span>(myTSSr, <span class="dt">method =</span> <span class="st">&quot;TPM&quot;</span>, <span class="dt">tpmLow =</span> <span class="fl">0.1</span>)</span>
<span id="cb7-9"><a href="#cb7-9"></a><span class="co">#&gt; </span></span>
<span id="cb7-10"><a href="#cb7-10"></a><span class="co">#&gt; Filtering data with TPM method...</span></span></code></pre></div>
<p>TSS clustering</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># TSS clustering </span></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="kw">clusterTSS</span>(myTSSr, <span class="dt">method =</span> <span class="st">&quot;peakclu&quot;</span>,<span class="dt">peakDistance=</span><span class="dv">100</span>,<span class="dt">extensionDistance=</span><span class="dv">30</span></span>
<span id="cb8-3"><a href="#cb8-3"></a>         ,<span class="dt">localThreshold =</span> <span class="fl">0.02</span>,<span class="dt">clusterThreshold =</span> <span class="dv">1</span></span>
<span id="cb8-4"><a href="#cb8-4"></a>         ,<span class="dt">useMultiCore=</span><span class="ot">FALSE</span>, <span class="dt">numCores=</span><span class="ot">NULL</span>)</span>
<span id="cb8-5"><a href="#cb8-5"></a></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="co"># Aggregating consensus clusters</span></span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="kw">consensusCluster</span>(myTSSr, <span class="dt">dis =</span> <span class="dv">50</span>, <span class="dt">useMultiCore =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>Core promoter shape</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co"># Calculating core promoter shape score</span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="kw">shapeCluster</span>(myTSSr,<span class="dt">clusters =</span> <span class="st">&quot;consensusClusters&quot;</span>, <span class="dt">method =</span> <span class="st">&quot;PSS&quot;</span>,</span>
<span id="cb9-3"><a href="#cb9-3"></a>             <span class="dt">useMultiCore=</span> <span class="ot">FALSE</span>, <span class="dt">numCores =</span> <span class="ot">NULL</span>)</span></code></pre></div>
<p>Annotation core promoters</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="co"># Assign clusters to the annotated features</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="kw">annotateCluster</span>(myTSSr,<span class="dt">clusters =</span> <span class="st">&quot;consensusClusters&quot;</span>,<span class="dt">filterCluster =</span> <span class="ot">TRUE</span>,</span>
<span id="cb10-3"><a href="#cb10-3"></a>              <span class="dt">filterClusterThreshold =</span> <span class="fl">0.02</span>, <span class="dt">annotationType =</span> <span class="st">&quot;genes&quot;</span></span>
<span id="cb10-4"><a href="#cb10-4"></a>              ,<span class="dt">upstream=</span><span class="dv">1000</span>, <span class="dt">upstreamOverlap =</span> <span class="dv">500</span>, <span class="dt">downstream =</span> <span class="dv">0</span>)</span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="co">#&gt; </span></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="co">#&gt; Annotating...</span></span></code></pre></div>
<p>Differential expression analysis</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="co"># Assign clusters to the annotated features</span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="kw">deGene</span>(myTSSr,<span class="dt">comparePairs=</span><span class="kw">list</span>(<span class="kw">c</span>(<span class="st">&quot;control&quot;</span>,<span class="st">&quot;treat&quot;</span>)), </span>
<span id="cb11-3"><a href="#cb11-3"></a>       <span class="dt">pval =</span> <span class="fl">0.01</span>,<span class="dt">useMultiCore=</span><span class="ot">FALSE</span>, <span class="dt">numCores=</span><span class="ot">NULL</span>)</span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="co">#&gt; </span></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="co">#&gt; Calculating gene differential expression...</span></span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="co">#&gt; converting counts to integer mode</span></span>
<span id="cb11-7"><a href="#cb11-7"></a><span class="co">#&gt; estimating size factors</span></span>
<span id="cb11-8"><a href="#cb11-8"></a><span class="co">#&gt; estimating dispersions</span></span>
<span id="cb11-9"><a href="#cb11-9"></a><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span id="cb11-10"><a href="#cb11-10"></a><span class="co">#&gt; mean-dispersion relationship</span></span>
<span id="cb11-11"><a href="#cb11-11"></a><span class="co">#&gt; final dispersion estimates</span></span>
<span id="cb11-12"><a href="#cb11-12"></a><span class="co">#&gt; fitting model and testing</span></span></code></pre></div>
<p>Core promoter shifts</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co"># Calcuate core promoter shifts</span></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="kw">shiftPromoter</span>(myTSSr,<span class="dt">comparePairs=</span><span class="kw">list</span>(<span class="kw">c</span>(<span class="st">&quot;control&quot;</span>,<span class="st">&quot;treat&quot;</span>)), <span class="dt">pval =</span> <span class="fl">0.01</span>)</span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="co">#&gt; </span></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="co">#&gt; Calculating core promoter shifts...</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
